### PR TITLE
[DSE] Fixes bug caused by skipped read clobbers

### PR DESCRIPTION
**Pull request by vporpo**
The bug seems to be caused by the use of BatchAA after the IR has been modified.

Considering the reproducer in `read-clobber-skipped.ll`, the bug is caused by `isWriteAtEndOfFunction()` returning true instead of false when trying to eliminate `@llvm.memset.po0`.
The problem is that when it visits `%ret = load ptr, ptr %struct.byte.8`, `isReadClobber()` should return true, but `isRefSet(BatchAA.getModRefInfo())` actually returns false, so the read clobber is missed. As far as I understand the cause is that a mem user store of `llvm.memset` has already been erased from the IR, causing BatchAA to return the wrong data.

Replacing BatchAA with AA at this point fixes the issue.

### CODE REVIEW

**Comment by vporpo** [comment]
I had to revert the pre-commit test because `BatchAA.getModRefInfo()` had some non-deterministic behavior, and the test was failing on some targets.

**Review by david-xl — COMMENTED** [review]
It is the removed dead store (at %struct.byte.4) that invalidates the batchAA right?

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp [hunk 1]
```diff
@@ -1274,7 +1274,8 @@ struct DSEState {
       if (CB->onlyAccessesInaccessibleMemory())
         return false;
 
-    return isRefSet(BatchAA.getModRefInfo(UseInst, DefLoc));
+    AAQueryInfo TmpAAQI(AA, &EI);
```
Comment (alinas) [inline llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:1277]: This shouldn't be necessary. `AA.getModRefInfo(UseInst, DefLoc)` should work.

**Comment by vporpo** [comment]
> It is the removed dead store (at %struct.byte.4) that invalidates the batchAA right?

Yes.

File: llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp (hunk 1 above)
Comment (vporpo) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:1277]: Ah, nice, thanks!

Comment (vporpo) [reply llvm/lib/Transforms/Scalar/DeadStoreElimination.cpp:1277]: Is there a better way to call `getModeRefInfo()` and using `EI` for the query, without having to use the temporary `AAQueryInfo` ?

**Comment by fhahn** [comment]
DSE

> > It is the removed dead store (at %struct.byte.4) that invalidates the batchAA right?
> 
> Yes.

This is an interesting case, thanks for sharing! AFAIK `BatchAA` caches pairs of memory locations; removing the store itself shouldn't invalidate anything in the cache. 

I noticed that the `memset` doesn't get removed incorrectly deterministically; sometimes it gets removed, sometimes it does not, so I took a closer look at whats' going on.

When removing a store, we also remove any instructions that become dead after removal (reversibly), so we may remove an instruction that was used as memory location pointer in the cache. In most cases, this isn't a problem either, as the instruction has been removed and we are left with stale entries in the cache, which shouldn't be accessed though.

What seems to be happening in the case at hand is that we create a new GEP for the adjusted memset and in some cases it gets allocated at the same address as a previous GEP that we removed, but is still in the cache.

I experimented a bit to confirm this was indeed the issue and it looks like we should be able to keep using BatchAA, if we avoid deleting instructions that may be cached before the end of DSE, which I sketched here https://github.com/llvm/llvm-project/pull/83411

**Comment by vporpo** [comment]
> I experimented a bit to confirm this was indeed the issue and it looks like we should be able to keep using BatchAA, if we avoid deleting instructions that may be cached before the end of DSE, which I sketched here https://github.com/llvm/llvm-project/pull/83411

Thanks for looking into it @fhahn and for figuring out how to fix it without resorting to using AA instead of BatchAA.

**Comment by vporpo** [comment]
Closing, the bug will be fixed by this patch: https://github.com/llvm/llvm-project/pull/83411

**Pull request closed without merging**